rust/lance-index/src/vector/ivf/storage.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use arrow_array::{Array, FixedSizeListArray, Float32Array};
+
+use lance_arrow::FixedSizeListArrayExt;
+use lance_core::{Error, Result};
+use lance_file::{reader::FileReader, writer::FileWriter};
+use lance_io::{traits::WriteExt, utils::read_message};
+use lance_table::io::manifest::ManifestDescribing;
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::{location, Location};
+
+use crate::pb::Ivf as PbIvf;
+
+const IVF_METADATA_KEY: &str = "lance:ivf";
+
+#[warn(dead_code)]
+#[derive(Debug, PartialEq)]
+pub struct IvfMetadadta {
+    /// Centroids of the IVF indices. Can be empty.
+    centroids: Arc<FixedSizeListArray>,
+
+    /// Length of each partition.
+    lengths: Vec<u32>,
+
+    /// pre-computed row offset for each partition, do not persist.
+    partition_row_offsets: Vec<usize>,
+}
+
+/// The IVF metadata stored in the Lance Schema
+#[derive(Serialize, Deserialize, Debug)]
+struct IvfMetadataInSchema {
```

Comment:
It's a little confusing to me that there is a "metadata" and a "metadata in schema".  I would think that "metadata" always belongs in the schema.  Maybe:

```
IvfMetadata -> IvfData
IvfMetadataInSchema -> IvfMetadata
```
?

Reply:
Sg